File: solutions/800.letter-case-permutation/letter-case-permutation.java
class Solution {
    public List<String> letterCasePermutation(String S) {
        List<String> res = new ArrayList<>(Arrays.asList(S));
        for(int i = 0; i < S.length(); i++) {
            for(int j = res.size() - 1; j >= 0 && Character.isLetter(S.charAt(i)); j--) {
                char[] ch = res.get(j).toCharArray();
                ch[i] ^= (1 << 5);
                res.add(String.valueOf(ch));
            }
        }
        return res;
    }
}

File: solutions/800.letter-case-permutation/letter-case-permutation.cpp
class Solution {
public:
    vector<string> letterCasePermutation(string S) {
        vector<string> res;
        dfs(S, 0, res);
        return res;
    }
private: 
    void dfs(string& S, int count, vector<string>& res) {
        if(S.size() == count) {
            res.push_back(S);
            return;
        }
        dfs(S, count + 1, res);
        if(isalpha(S[count])) {
            S[count] ^= (1 << 5);
            dfs(S, count + 1, res);
        }
    }
};